### PR TITLE
System Utils Package

### PR DESCRIPTION
**Pull request by kyy2**
Part of #247

### CODE REVIEW

File: core/catkin_ws/src/custom_msgs/msg/Memory.msg [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Common interface for memory items (RAM, disk, GPU memory)
+
+float64 used
+float64 total
+float64 percentage
```
Comment (muthuArivoli) [inline core/catkin_ws/src/custom_msgs/msg/Memory.msg:5]: newline at end

File: core/catkin_ws/src/custom_msgs/msg/SystemUsage.msg [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+float64 cpu_percent
+float64 cpu_speed
+
+float64 gpu_percent
+float64 gpu_speed
+Memory gpu_memory
+
+Memory ram
+
+Memory disk
```
Comment (muthuArivoli) [inline core/catkin_ws/src/custom_msgs/msg/SystemUsage.msg:14]: newline at end

File: docker/onboard/Dockerfile [hunk 3]
```diff
@@ -35,6 +35,11 @@ RUN ./TorchInstaller.sh && \
 # Install Detecto
 RUN pip install --no-cache-dir --no-deps detecto
 
+# Install psutils and GPUtil
+RUN pip install \
+    psutils \
+    gputil
+
```
Comment (muthuArivoli) [inline docker/onboard/Dockerfile:42]: include these in line 28 with pyserial and other pip.
also fix package name

File: onboard/catkin_ws/src/system_usage/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.0.2)
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/CMakeLists.txt:1]: rename system_utils

File: onboard/catkin_ws/src/system_usage/launch/pub_system.launch [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<launch>
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/launch/pub_system.launch:1]: rename launch file to system_pub.launch

File: onboard/catkin_ws/src/system_usage/package.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>system_usage</name>
+  <version>0.0.0</version>
+  <description>
+    Package that publishes system usage information
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/package.xml:6]: description to something more generic

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+class SystemInfoPublisher:
+    TOPIC_NAME = 'system/usage'
+    NODE_NAME = 'system_usage_publisher'
+
+    def __init__(self):
+        self._pub = rospy.Publisher(self.TOPIC_NAME, SystemUsage, queue_size=10)
+        self._current_msg = SystemUsage()
+        self.gpu_memory = Memory()
+        self.ram_memory = Memory()
+        self.disk_memory = Memory()
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:18]: don't use global variables if you don't have to

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    def get_gpu_memory_others(self):
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:20]: local to function

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            self._current_msg.gpu_percent = gpu.load * 100
+            self._current_msg.gpu_speed = 0
+        else:
+            self.gpu_memory.used = 0
+            self.gpu_memory.total = 0
+            self.gpu_memory.percentage = 0
+        return self.gpu_memory
+
+    def get_ram_memory(self):
+        self.ram_memory.used = psutil.virtual_memory().total - psutil.virtual_memory().available
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:36]: return a msg

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        self.ram_memory.total = psutil.virtual_memory().total
+        self.ram_memory.percentage = psutil.virtual_memory().percent
+        return self.ram_memory
+
+    def get_disk_memory(self):
+        self.disk_memory.used = psutil.disk_usage('/').used
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:42]: return a msg

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        self.disk_memory.total = psutil.disk_usage('/').total * 0.95
+        self.disk_memory.percentage = psutil.disk_usage('/').percent
+        return self.disk_memory
+
+    def run(self):
+        rospy.init_node(self.NODE_NAME)
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:48]: move init_node to init

File: onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            self._current_msg.ram = ram
+
+            self._current_msg.disk = disk
+            self._current_msg.cpu_percent = psutil.cpu_percent(interval=0.5)
+            self._current_msg.cpu_speed = psutil.cpu_freq().current
+
+            self._current_msg.gpu_memory = gpu
+
+            self._current_msg.header.stamp = rospy.Time.now()
+            self._current_msg.header.frame_id = "system_usage_link"
```
Comment (muthuArivoli) [inline onboard/catkin_ws/src/system_usage/scripts/system_info_publisher.py:63]: remove frame_id

**Review by muthuArivoli — CHANGES_REQUESTED** [review]
Some changes

**Review by ShaanGondalia — APPROVED** [review]
LGTM

**Review by muthuArivoli — APPROVED** [review]
LGTM

**Pull request merged**